doc/bptdump.1
.\" Man page for bptdump
.\" Use the following command to view man page:
.\"
.\"  tbl bptdump.1 | nroff -man | less
.\"
.TH BPTDUMP 1 "July 22, 2012"
.SH NAME
bptdump \- A utility to dump an Emdros database as a Bit Packed Table
.SH SYNOPSIS
\fBbptdump\fR [ options ] -d database-to-dump bptfilename
.br
\fBbptdump\fR [ options ] 
.br
.SH DESCRIPTION
\fBbptdump\fR is a utility to dump an Emdros database created with one
of the "other" backends as a BPT (Bit Packed Table) database, suitable
for the BPT engine.
.TP
The BPT engine is a lightweight backend database engine that is
tailored to the EMdF model which Emdros implements.  It is much faster
than the other backends, and creates databases that are smaller than
the other backends.  It is proprietary, not Open Source.
.TP
Please contact Ulrik Sandborg-Petersen, the author of Emdros, for more
information about the BPT engine, including licensing options.  You
may reach him via: <ulrikp{at}emergence[dot]dk> (address obfuscated to
avoid spammers; sorry to everyone else...).
.TP
In the Open Source version of Emdros, the \fBbptdump\fR utility does
nothing beyond report an error.  The program only does something
useful in the proprietary version.
.TP
The BPT engine has the ability to have, within the same file, two BPT
databases: One which is the "main" database, and one which is a
"payload", i.e., a compressed version of a secondary database.  The
payload cannot be read as part of the main database, but it can be
dumped into another file, using on-the-fly encryption if need be,
using the method BPTEMdFDB::dumpPayload().

.SH OPTIONS
\fBbptdump\fR supports the following command-line switches:
.TP 6
.BI \-\-help
show help
.TP
.BI \-V\ ,\ \-\-version
show version
.TP
.BI \-d\ ,\ \-\-dbname \ dbname 
set name of the database to dump.  
.TP
.BI \--key\ key
set the key with which to encrypt the BPT database
.TP
.BI \-pd\ ,\ \-\-payload\-db \ dbname 
set name of any payload database to compress and add to the end
.TP
.BI \-pk\ ,\ \-\⨪payload\-key \ key
set the key with which to decrypt the payload before compressing
.TP
.BI \--schema\ (1|3)
Use the specified BPT schema. Schema version 1 is a row-store. Schema
version 3 is a column-store. The default is 3. Version 2 of the schema
is a variant of version 1, and is used automatically. It is only used
if the database to be dumped has at least one feature whose type is
SET OF MONADS. It is not possible to specify that version 2 must be
used.
.TP
.BI \-h\ ,\ \-\-host \ hostname 
set db back-end hostname to connect to (default is 'localhost').  Has
no effect on SQLite.
.TP
.BI \-u\ ,\ \-\-user \ user
set database user to connect as (default is 'emdf').  Has no effect on
SQLite.
.TP
.BI \-p\ ,\ \-\-password \ password
set password to use for the database user.  Has no effect on SQLite,
unless you have an encryption-enabled SQLite, in which case this gets
passed as the key.
.TP
.BI \-b\ ,\ \-\-backend \ backend
set database backend to `backend'. Valid values are: For PostgreSQL:
"p", "pg", "postgres", and "postgresql". For MySQL: "m", "my", and
"mysql". For SQLite 2.X.X: "2", "s", "l", "lt", "sqlite", and
"sqlite2". For SQLite 3.X.X: "3", "s3", "lt3", and "sqlite3".

.SH EXAMPLES

The following would dump a PostgreSQL database (-b p) with the
PostgreSQL username "mike", mike's password being "sicrit", from the
host "other.host.mydomain.net", and dumping the PostgreSQL database
mydbtodump into the BPT file /home/mike/bptoutputfilename.
.TP
bptdump -b p -u mike -p sicrit -h other.host.mydomain.net -d mydbtodump /home/mike/bptoutputfilename
.TP
Another example, this time dumping an SQLite 3 database called
"/home/mike/db/mydbtodump", and dumping it into /home/mike/db/bptdb.
.TP
bptdump -b 3 -d /home/mike/db/mydbtodump /home/mike/db/bptdb


.SH RETURN VALUES
.TP
.BR 0 " Success"
.TP
.BR 1 " Wrong usage"
.TP
.BR 2 " Connection to backend server could not be established"
.TP
.BR 3 " An exception occurred (the type is printed on stderr)"
.TP
.BR 4 " Something else went wrong."
.TP
.SH AUTHORS
Copyright
.Cr
2001-2012 by Ulrik Petersen (ulrikp@users.sourceforge.net).


NEWS
This file is for release-notes for Emdros.

- *** Version 3.5.0 *** 

Release notes for Emdros 3.5.0
==============================

The focus-areas of this release include:

- Enhancements to the core MQL
- Bugfixes to the core MQL.
- Enhancements to the render library
- Enhancements to the Full Text Search subsystem
- Enhancements to SWIG bindings
- Enhancements to the build system

Please consult the ChangeLog (going back to version 3.4.1.pre01) for
the specifics.  You can also consult the MQL Programmer's Guide for
the enhancements to the core MQL, and the Programmer's Reference Guide
for the enhancements to the Render library and the FTS subsystem.

Briefly, the core MQL has been enhanced as follows:

- SELECT ALL OBJECTS has a new WITH MAX RANGE clause, helping to limit
  otherwise runaway queries such as [Word]..[Word].

- SELECT ALL OBJECTS now sports an [UnorderedGroup] block.

- GET AGGREGATE FEAUTURES is a new statement, similar to SQL's MIN(),
  MAX(), SUM(), and COUNT() aggregate functions.


- *** Version 3.4.0 *** 

Release notes for Emdros 3.4.0
==============================

The focus-areas of this release include:

- Support for iOS and Android
- A smoother user experience for the GUI applications
- Enhancements to the MQL query language
- Documentation improvements
- Build improvements on all platforms
- Speed improvements


Binaries for Windows(R) and Mac OS X(R)
---------------------------------------

As usual, binaries are available for Mac OS X and Windows(R).

The Windows binaries are available in both x86 and x64 versions.  They
have built-in support for MySQL, SQLite 2, and SQLite 3.  The x86
binaries are built with Visual Studio Express 2010.  The x64 binaries
are built with Windows SDK 6.0A.

The Mac OS X binaries are Universal binaries (i386 and x86_64) running
on Mac OS X 10.6 (Snow Leopard), 10.7 (Lion), and 10.8 (Mountain
Lion), 10.9 (Mavericks), and possibly later versions of Mac OS X.
They do not have support for either MySQL or PostgreSQL; Only SQLite 2
and SQLite 3 are supported in the Mac OS X binaries.  You might be
able to compile the sources with support for MySQL yourself, and
possibly also PostgreSQL.

Linux binaries
--------------

For this release, we do not provide Linux binaries.  Instead, we have
ensured that it is easy to build .rpms and .debs from the sources.


To build an RPM or .deb, first install the dependencies mentioned in
the text file called "INSTALL" in the root of the sources.


A) To build an RPM:

rpmbuild -ta emdros-3.4.0.tar.gz


B) To build a .deb:

1) cp /path/to/Emdros/sources/emdros-3.4.0.tar.gz .

2) tar xfvz emdros-3.4.0.tar.gz

3) cd emdros-3.4.0

4) dpkg-buildpackage -rfakeroot

This will build the .deb in ../.


Changes in more detail
----------------------

Release 3.4.0 has the following changes over 3.3.0:

- Build changes:

  - Support for compilation on Android and iOS were added.

  - Non-Unicode (ANSI) builds of the programs are no longer available
    on Windows.

  - On Linux / Unix / Mac OS X, wx/libemdrosguiu is now a static
    library.

  - Out-of-the-box support for building on Debian / Ubuntu /
    derivatives was added:

	$ tar xfzv emdros-x.y.z.tar.gz
	$ cd emdros-x.y.z
	$ dpkg-buildpackage -rfakeroot
	$ cd .. # The .deb package is in this directory
	$ sudo dpkg -i emdros_*.deb # This should install it.

  - The .pro build files for Qt have been updated to Qt 4.8.0, which
    entails dropping support for everything below Mac OS X 10.6 on Qt.

  - It is now possible to do Win64 builds on Windows.  See
    the very top of win32\config.mak for how to do it.

  - It is now possible to build Emdros on Windows with MinGW32,
    including support for wxWidgets.

  - Add a test suite which is run with "make test" in the root
    directory of the sources.

  - Much better configure.ac support for wxWidgets 2.8, 2.9, and 3.0,
    especially on Mac OS X.


- Harvest library changes:

  - Bugs were fixed
  - Speeds were increased
  - The template language was enhanced (see the docs)
  - A new Full Text Search engine, FTS2, was added.

- PHP was added as a supported SWIG binding.

- More support for feature-type "SET OF MONADS" was added:

  - Bugs were fixed
  - In topographic queries, it is now possible to say:

    SELECT ALL OBJECTS WHERE 
    [Clause 
        [Phrase
           // Could also be universe instead of substrate.
           // Could also be part_of instead of overlap.
           // part_of(monads, substrate) is still the default.
           overlap(mymonadsetfeature, substrate)
        ]
    ]

- The GET OBJECTS HAVING MONADS IN statement now supports using a
  different monad set than the default "monads" feature:

  GET OBJECTS HAVING MONADS IN
  { 2-40201, 40203-50000 }
  USING MONAD FEATURE described_SOM
  [Describer]
  GO

- Assorted speed increases in the EMdF / MQL libraries.

- The Penn Treebank converter output was improved.

- SQLite 3 was updated to version 3.8.4.3.

- Fixed a few documentation glitches


Enjoy!


Ulrik Sandborg-Petersen


- *** Version 3.3.0 *** 

As usual, binaries are available for Mac OS X, Windows(R), and Fedora
(15).  

The Windows binaries have support for MySQL, SQLite 2, and SQLite 3.
They are built with Visual Studio Express 2008.

The Mac OS X binaries are Universal binaries running on Mac OS X 10.4
(Tiger), 10.5 (Leopard), and 10.6 (Snow Leopard).  They do not have
support for either MySQL or PostgreSQL; Only SQLite 2 and SQLite 3 are
supported in the Mac OS X binaries.  You can compile the sources with
support for MySQL yourself, though, and possibly also PostgreSQL.

The Fedora binaries come with support for PostgreSQL, MySQL, SQLite 2,
and SQLite 3.

Release 3.3.0 has the following changes over 3.2.0:


src/bptdump.cpp
/*
 * bptdump.cpp
 *
 * A front-end for dumping an Emdros database as a Bit Packed Table
 * (BPT)
 *
 * Created: 5/1-2001 (1st of May, 2001)
 * Last update: 4/3-2018
 *
 * Return codes:
 * 
 * 0 : Success
 * 1 : Wrong usage
 * 2 : Connection to backend server could not be established
 * 3 : An exception occurred
 * 4 : Something else went wrong
 *
 */
/************************************************************************
 *
 *   Emdros - the database engine for analyzed or annotated text
 *   Copyright (C) 2001-2018  Ulrik Sandborg-Petersen
 *
 *   This program is free software; you can redistribute it and/or
 *   modify it under the terms of the GNU General Public License as
 *   published by the Free Software Foundation, license version 2.  
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *   General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the Free Software
 *   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
 *   02111-1307 USA
 *
 *
 *   Special exception
 *   =================
 * 
 *   In addition, as a special exception, Ulrik Petersen, the
 *   copyright holder of Emdros, gives permission to link Emdros, in
 *   whole or in part, with the libraries which are normally
 *   distributed with:
 *   
 *   a) Sun's Java platform,
 *   b) Python, 
 *   c) Jython,
 *   d) Ruby, and/or 
 *   e) Perl 
 *   f) PostgreSQL
 *   g) OpenSSL
 *
 *   (or with modified versions of these), and to distribute linked
 *   combinations including both Emdros, in whole or in part, and one
 *   or more of the libraries normally distributed with (a)-(g) above.
 *
 *   Please note: This gives you special rights concerning the
 *   libraries which normally accompany the above pieces of software.
 *   It gives you no special rights concerning software that you write
 *   yourself.  You must obey the GNU General Public License in all
 *   respects for all of the code used other than the libraries
 *   normally distributed with (a)-(g) above.
 *
 *   If you modify this file, you may extend this exception to your
 *   version of the file, but you are not obligated to do so. If you
 *   do not wish to do so, delete this exception statement from your
 *   version.
 *
 *
 *   Other licensing forms
 *   =====================
 *
 *   If you wish to negotiate commercial licensing, please contact
 *   Ulrik Petersen at ulrikp[at]users.sourceforge.net.
 *
 *   Licensing can also be negotiated if your organization is an
 *   educational, non-profit, charity, missionary or similar
 *   organization.
 *
 *
 *   Website
 *   =======
 *
 *   Emdros has a website here:
 *
 *   http://emdros.org
 *
 *
 *
 **************************************************************************/


#include <emdros-config.h>
// Do this because emdros-config might #define malloc,
// in which case some systems will fail to use the real 
// malloc, thus causing build errors.
#undef malloc


#include <monads.h>
#include <iostream>
#include <sstream>
#include <fstream>
#include <string_func.h>
#include <emdfdb.h>
#include <table.h>
#include <emdros-lconfig.h>
#include <emdf_output.h>
#include <emdros_environment.h>
#include <encryption.h>
#include <opt.h>
#include <messages.h>

#include <bptdumper.h>
#include <bpt2dumper.h>


bool exec_bptdumper(EmdrosEnv *pSourceEE, const std::string& bptfilename, const std::string& bptkey, eBackendKind backend_kind, long schema_version, const std::string& payload_filename, const std::string& payload_inkey, const std::string& payload_codec)
{
	bool bResult = false;

	std::string error_message;
	if (backend_kind == kBPT) {
		bResult = BPTdumpAsBPT(pSourceEE, bptfilename, bptkey, error_message, schema_version, payload_filename, payload_inkey, payload_codec);
	} else if (backend_kind == kBPT2) {
		bResult = BPT2dumpAsBPT2(pSourceEE, bptfilename, bptkey, error_message, schema_version, payload_filename, payload_inkey);
	} else {
		bResult = false;
		error_message = "Unknown BPT backend.\n\n";
	}
	if (bResult) {
		std::cout << "SUCCESS!\nThe database was successfully dumped as:\n" << bptfilename << std::endl;
	} else {
		std::cerr << "FAILURE!\nSomething went wrong while dumping database:\n\n"
			  << "---------------------------------------------------------------\n"
			  << error_message 
			  << "\n---------------------------------------------------------------\n"
			  << "FAILURE!" << std::endl;
			
	}

	// Return result
	return bResult;
}



void show_version(std::ostream& ostr)
{
	ostr << "bptdumper from Emdros version " << EMDROS_VERSION << std::endl;
}

void print_usage(std::ostream& ostr)
{
	show_version(ostr);
	ostr << "Usage:" << std::endl;
	ostr << "   bptdumper [options] bptfilename " << std::endl;
	ostr << "   bptdumper (--help|-V|--version)" << std::endl;
	ostr << "OPTIONS:" << std::endl;
	printUsageStandardArguments(ostr);
	ostr << "   -d , --dbname db     Dump this database (required)" << std::endl;
	ostr << "   -pd, --payload-db db    Add this database as a payload (default none)" << std::endl;
	ostr << "   -pk, --payload-key key  Use this key to decrypt the payload (default none)" << std::endl;
	ostr << "   -pc, --payload-codec cd Use this code to compress the payload (zlib, lzma)" << std::endl;
	ostr << "   --key key               Encrypt BPT db with this key" << std::endl;
	ostr << "   -bv, --bpt-version (1|2) Which BPT version to use" << std::endl;
	ostr << "   --schema (1|3|5|7)      Which BPT schema version to use" << std::endl;
	ostr << "DEFAULTS:" << std::endl;
	printUsageDefaultsOfStandardArguments(ostr);
	ostr << "   --bpt-version 1" << std::endl;
	ostr << "   --schema 3 (for BPT version 1)" << std::endl;
	ostr << "   --schema 1 (for BPT version 2)" << std::endl;
	ostr << "   --payload-codec zlib" << std::endl;
}

int main(int argc, char* argv[])
{
	// Set defaults
	eOutputKind output_kind = kOKConsole;
	std::string db_to_dump;
	std::string bptkey;
	std::string bptfilename;
	std::string bptschema_version;
	std::string bpt_version;
	long schema_version;
	std::string hostname;
	std::string user;
	std::string password;
	eBackendKind backend_kind;
	eCharsets charset;
	std::string payload_filename;
	std::string payload_inkey;
	std::string payload_codec = "zlib";
	eBackendKind bpt_backend_kind = kBPT;


	addStandardArguments(false // Do add -e | --encoding
			     );

	addOption("-d", "--dbname", true, 
		  "",
		  "ERROR: -d and --dbname must have a database-name as their argument.\n"
		  "       example: -d book1\n");

	addOption("--key", "--key", true, 
		  "",
		  "ERROR: --key must have a key as its argument.");

	addOption("-pd", "--payload-db", true, 
		  "",
		  "ERROR: -pd and --payload-db must have a BPT database-name as their argument.\n"
		  "       example: -pd book1\n");

	addOption("-pk", "--payload-key", true, 
		  "",
		  "ERROR: --pk and --payload-key must have a key as its argument.");

	addOption("-pc", "--payload-codec", true, 
		  "zlib",
		  "ERROR: -pc and --payload-codec must have either 'zlib' or 'lzma' as its argument.");

	addOption("--schema", "--schema", true, 
		  "3",
		  "ERROR: --key must have a key as its argument.");

	addOption("-bv", "--bpt-version", true, 
		  "1",
		  "ERROR: --key must have a key as its argument.");

	// Parse arguments
	std::list<std::string> surplus_arguments;

	std::string error_message;
	if (!parseArguments(argc, argv, error_message, surplus_arguments)) {
		print_usage(std::cerr);
		std::cerr << error_message << std::endl;
		return 1;
	} else {
		bool bShowVersion;
		bool bShowHelp;
		std::string error_message;
		if (!getStandardArguments(bShowVersion, bShowHelp,
					  hostname,
					  user,
					  password,
					  backend_kind,
					  charset,
					  error_message)) {
			print_usage(std::cerr);
			std::cerr << error_message << std::endl;
			return 1;
		}

		// -V | --version
		if (bShowVersion) {
			show_version(std::cout);
			return 0;
		}

		// --help
		if (bShowHelp) {
			print_usage(std::cout);
			return 0;
		}

		// Database name
		getArgumentValue("-d", db_to_dump);

		// Key
		getArgumentValue("--key", bptkey);

		// Payload db
		getArgumentValue("-pd", payload_filename);

		// Payload key
		getArgumentValue("-pk", payload_inkey);

		// Payload codec
		getArgumentValue("-pc", payload_codec);

		// Schema
		getArgumentValue("--schema", bptschema_version);

		// Payload db
		getArgumentValue("-bv", bpt_version);

		if (bpt_version == "1") {
			schema_version = string2long(bptschema_version);
			bpt_backend_kind = kBPT;
		} else if (bpt_version == "2") {
			// We don't have any other than schema version 1
			schema_version = 1;
			bpt_backend_kind = kBPT2;
		} else {
			std::cerr << "Error: Unknown bpt backend version: " << bpt_version << ".\n"
				  << "       Only known versions so far: 1 and 2.\n" << std::endl;
			return 1;
		}

		if (payload_codec != "zlib"
		    && payload_codec != "lzma") {
			std::cerr << "Error: Unknown bpt payload codec: " << payload_codec << std::endl;
			std::cerr << "       Currently known codecs: 'zlib' and 'lzma'." << std::endl;
			return 1;
		}

		if (surplus_arguments.size() == 1) {
			bptfilename = *surplus_arguments.begin();
		} else {
			std::cerr << "Error: You can only specify one bptfilename. You specified: '" << joinList("' and '", surplus_arguments) << "'." << std::endl;
			return 1;
		}
	}

	// Make connection
	EmdrosEnv *pEE = new EmdrosEnv(output_kind, 
				       charset,
				       hostname,
				       user, password,
				       db_to_dump,
				       backend_kind);
	// Zero-fill password
	zeroFillString(password);

	// Check that we connected
	if (!pEE->connectionOk()) {
		emdrosMessageConnectionNotOK(&std::cerr, backend_kind);
		delete pEE;
		return 2;
	}

	int nResult = 3;
	try {
		if (exec_bptdumper(pEE, bptfilename, bptkey, bpt_backend_kind, schema_version, payload_filename, payload_inkey, payload_codec)) {
			nResult = 0;
		} else {
			nResult = 4;
		}
	} catch (EMdFNULLValueException e) {
		std::cerr << "ERROR: EMdFNULLValueException (Database error)..." << std::endl;
		std::cerr << pEE->getDBError() << std::endl; 
		std::cerr << pEE->getCompilerError() << std::endl;
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EMdFDBDBError e) {
		std::cerr << "ERROR: EMdFDBDBError (Database error)..." << std::endl;
		std::cerr << pEE->getDBError() << std::endl; 
		std::cerr << pEE->getCompilerError() << std::endl;
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EMdFDBException e) {
		std::cerr << "ERROR: EMdFDBException (Database error)..." << std::endl;
		std::cerr << pEE->getDBError() << std::endl; 
		std::cerr << pEE->getCompilerError() << std::endl;
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (TableColumnException e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (TableRowException e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (TableException e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (BadMonadsException e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (WrongCharacterSetException e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EMdFOutputException e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EmdrosException e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (...) {
		std::cerr << "Unknown exception occurred.  Program aborted." << std::endl;
	} 

	// Clean up
	delete pEE;
	return nResult;
}
